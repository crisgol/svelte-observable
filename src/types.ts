export type Unsubscribe = () => void;
export type Next<T> = (value: T) => void;

export interface Subscriber<T> {
  next?: Next<T>;
  error?(error: Error): void;
  complete?(): void;
}
export interface Subscription {
  unsubscribe: Unsubscribe;
}

export interface Observable<T> {
  subscribe(subscriber: Next<T> | Subscriber<T>): Subscription;
}

export interface ReadableStore<T> {
  subscribe(subscriber: Next<T>): Unsubscribe;
}

export type Fulfill<T> = (value: T) => void;
export type Reject = (error: Error) => void;
export type Deferred<T> = T | Promise<T>;
