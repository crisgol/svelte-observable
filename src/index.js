export { default as observe } from './observe';
export { default as subscribe } from './subscribe';
export { default as unsubscribe } from './unsubscribe';
