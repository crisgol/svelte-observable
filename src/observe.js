import { readable } from 'svelte/store';
import { isObservable, deferred } from './utils';

const noop = () => {};

export default function observe(observable, initial) {
  if (!isObservable(observable)) {
    return readable(noop, observable);
  }

  return readable(set => {
    const { fulfill, reject } = deferred(set, initial);

    const subscription = observable.subscribe({
      next(value) {
        fulfill(value);
      },
      error(err) {
        reject(err);
      }
    });

    return () => subscription.unsubscribe();
  });
}
