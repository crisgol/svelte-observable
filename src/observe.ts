import { readable } from 'svelte/store';
import { isObservable, deferred } from './utils';
import { Observable, ReadableStore, Deferred } from './types';

const noop = () => {};

export default function observe<T>(
  observable: any,
  initial?: T
): ReadableStore<Deferred<T>> {
  if (!isObservable(observable)) {
    return readable(observable as T, noop);
  }

  return readable((undefined as unknown) as Deferred<T>, set => {
    const { fulfill, reject } = deferred<T>(set, initial);

    const subscription = (observable as Observable<T>).subscribe({
      next(value) {
        fulfill(value);
      },
      error(err) {
        reject(err);
      }
    });

    return () => subscription.unsubscribe();
  });
}
