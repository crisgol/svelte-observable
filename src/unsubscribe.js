import { SVELTE_OBSERVABLE } from './utils';

export default function unsubscribe(component) {
  component = component || this;

  const subscriptions = component[SVELTE_OBSERVABLE];
  if (!subscriptions) return;

  for (const key in subscriptions) {
    subscriptions[key].unsubscribe();
  }

  component[SVELTE_OBSERVABLE] = null;
}
