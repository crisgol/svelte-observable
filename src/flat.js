import { readable } from 'svelte/store';
import { isObservable, pending, fulfilled, rejected } from './utils';
import observe from './observe';

export default function flat(subscribable) {
  const is_observable = isObservable(subscribable);

  return readable(
    set => {
      let inner_unsubscribe;
      let outer_unsubscribe;

      const fulfill = is_observable
        ? value => set(fulfilled(value))
        : value => set(value);

      function next(value) {
        if (inner_unsubscribe) inner_unsubscribe();
        if (isObservable(value)) value = observe(value);

        if (isSubscribable(value)) {
          inner_unsubscribe = value.subscribe(inner => fulfill(inner));
        } else {
          fulfill(value);
        }
      }
      function error(err) {
        set(rejected(err));
      }

      if (is_observable) {
        const subscription = subscribable.subscribe({ next, error });
        outer_unsubscribe = () => subscription.unsubscribe();
      } else {
        outer_unsubscribe = subscribable.subscribe(next);
      }

      return () => {
        if (inner_unsubscribe) inner_unsubscribe();
        outer_unsubscribe();
      };
    },
    is_observable ? pending() : undefined
  );
}

function isSubscribable(value) {
  return value && typeof value.subscribe === 'function';
}
