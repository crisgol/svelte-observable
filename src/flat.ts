import { readable } from 'svelte/store';
import { isObservable, deferred } from './utils';
import observe from './observe';
import {
  ReadableStore,
  Unsubscribe,
  Observable,
  Fulfill,
  Reject,
  Deferred
} from './types';

type Store<T> = ReadableStore<T> & object;
type Subscribable<T> = Observable<T> | Store<T>;
type Any<T> = T | Observable<T> | Store<T>;

export default function flat<T>(
  subscribable: Subscribable<Any<T>>,
  initial?: T
): ReadableStore<Deferred<T>> {
  const is_observable = isObservable(subscribable);

  return readable((undefined as unknown) as Deferred<T>, set => {
    let inner_unsubscribe: Unsubscribe | null = null;
    let outer_unsubscribe: Unsubscribe | null = null;

    const { fulfill = (value: T) => set(value), reject } = (is_observable
      ? deferred<T>(set, initial)
      : {}) as { fulfill?: Fulfill<T>; reject?: Reject };

    function next(value: Any<T>) {
      if (inner_unsubscribe) {
        inner_unsubscribe();
        inner_unsubscribe = null;
      }
      if (isObservable(value))
        value = observe(value as Observable<T>) as Store<T>;

      if (isStore(value)) {
        inner_unsubscribe = value.subscribe(inner => fulfill(inner));
      } else {
        fulfill(value as T);
      }
    }
    function error(error: Error) {
      reject!(error);
    }

    if (is_observable) {
      const subscription = (subscribable as Observable<Any<T>>).subscribe({
        next,
        error
      });
      outer_unsubscribe = () => subscription.unsubscribe();
    } else {
      outer_unsubscribe = (subscribable as Store<Any<T>>).subscribe(next);
    }

    return () => {
      if (inner_unsubscribe) inner_unsubscribe();
      outer_unsubscribe!();
    };
  });
}

function isStore(value: any): value is Store<any> {
  return value && typeof value.subscribe === 'function';
}
