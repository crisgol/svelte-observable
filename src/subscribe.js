import { SVELTE_OBSERVABLE, nonenumerable } from './utils';
import unsubscribe from './unsubscribe';

export default function subscribe({ changed, current, previous }) {
  // Load subscriptions from component
  // (if none found, assume first run and attach destroy listener to unsubscribe)
  let subscriptions = this[SVELTE_OBSERVABLE];
  if (!subscriptions) {
    nonenumerable(this, SVELTE_OBSERVABLE, {});
    subscriptions = this[SVELTE_OBSERVABLE];

    this.on('destroy', () => unsubscribe(this));
  }

  // Check for observables and subscribe / unsubscribe as-needed
  for (const key in changed) {
    const value = current[key];
    const observable = value && value[SVELTE_OBSERVABLE];
    const previous_observable =
      previous && previous[key] && previous[key][SVELTE_OBSERVABLE];

    if (observable && observable === previous_observable) {
      continue;
    }

    const existing = subscriptions[key];
    if (existing) {
      // Observable was overwritten, unsubscribe
      existing.unsubscribe();
    }

    if (observable) {
      subscriptions[key] = createSubscription(this, key, value);
    }
  }
}

function createSubscription(component, key, observer) {
  // Subscribe to changes
  // - First change, resolve initial deferred value
  // - Subsequent changes, set on component
  const observable = observer[SVELTE_OBSERVABLE];
  let resolved = false;

  const replace = value => {
    if (!resolved && observer.resolve) {
      resolved = true;
      observer.resolve(value);
    } else {
      // Continue exposing underlying observable on set values
      nonenumerable(value, SVELTE_OBSERVABLE, observable);
      component.set({ [key]: value });
    }
  };

  const subscription = observable.subscribe({
    next: value => replace(Promise.resolve(value)),
    error: error => replace(Promise.reject(error))
  });

  return subscription;
}
