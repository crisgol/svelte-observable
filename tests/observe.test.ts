import Observable, { Observer } from 'zen-observable';
import { load, check, Result, Rejected } from './helpers';
import { observe } from '../src';

it('should have fulfilled values', async () => {
  const observable = Observable.of(1, 2, 3);
  const store = observe(observable);

  const values = await load(store);
  const states = await check(values);

  expect(states).toMatchSnapshot();
});

it('should have rejected error values', async () => {
  const observable = new Observable((observer: Observer) => {
    observer.error(new Error('Uh oh.'));
  });
  const store = observe(observable);

  const values = await load(store);
  const states = (await check(values)) as Result<any>[];

  expect((states[0] as Rejected).rejected).toBe(true);
  expect((states[0] as Rejected).error.message).toBe('Uh oh.');
});

it('should have readable passthrough if not Observable', async () => {
  const store = observe(4);

  const values = await load(store);
  expect(values[0]).toBe(4);
});

it('should have initial value', async () => {
  const observable = Observable.of(1, 2, 3);
  const store = observe(observable, 0);

  const values = await load(store);
  const states = await check(values);

  expect(states).toMatchSnapshot();
});
