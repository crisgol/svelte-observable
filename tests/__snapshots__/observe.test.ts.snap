// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have fulfilled values 1`] = `
Array [
  Object {
    "fulfilled": true,
    "value": 1,
  },
  Object {
    "fulfilled": true,
    "value": 2,
  },
  Object {
    "fulfilled": true,
    "value": 3,
  },
]
`;

exports[`should have initial value 1`] = `
Array [
  0,
  Object {
    "fulfilled": true,
    "value": 1,
  },
  Object {
    "fulfilled": true,
    "value": 2,
  },
  Object {
    "fulfilled": true,
    "value": 3,
  },
]
`;
